File: src/prototype/web/src/state/stamina.ts
// PURE — stamina update per docs/design/state_machines_v1.md §5.
//
// Model: single scalar [0,1]. Per-second rates sum, then integrate over dt.
// Consumption and recovery rates are design-constant; §5.4 notes that M1
// implements consumption + recovery + threshold clamps but not visuals.

import { breakMagnitude } from "./posture_break.js";
import type { ActorState } from "./game_state.js";
import type { HipIntent } from "../input/intent.js";

export type StaminaConfig = Readonly<{
  // §5.2
  handActiveDrainPerSec: number;     // any hand REACHING / GRIPPED(≥0.5)
  postureMaintainDrainPerSec: number; // × ‖posture_break‖
  confirmDrainFlat: number;           // applied once per confirm via applyConfirmCost
  breathRecoverPerSec: number;        // BTN_BREATH + no grips + static
  idleRecoverPerSec: number;          // all IDLE, no grips

  hipInputStaticThreshold: number;    // §5.2 "hip_input < 0.2"

  // §5.3
  lowGripCap: number;                  // grip strength clamp at low stamina
  lowGripCapThreshold: number;         // stamina level below which the cap activates
  noReachThreshold: number;            // stamina level below which REACHING is refused
}>;

export const DEFAULT_STAMINA_CONFIG: StaminaConfig = Object.freeze({
  handActiveDrainPerSec: 0.02,
  postureMaintainDrainPerSec: 0.05,
  confirmDrainFlat: 0.1,
  breathRecoverPerSec: 0.1,
  idleRecoverPerSec: 0.03,
  hipInputStaticThreshold: 0.2,
  lowGripCap: 0.6,
  lowGripCapThreshold: 0.2,
  noReachThreshold: 0.05,
});

export type StaminaInputs = Readonly<{
  dtMs: number;
  actor: ActorState;
  attackerHip: HipIntent;
  triggerL: number;
  triggerR: number;
  breathPressed: boolean;   // BTN_BREATH held or edge this frame (treat as "trying to breathe")
}>;

// Classification helpers ----------------------------------------------------

function handIsActive(handState: string, triggerValue: number): boolean {
  if (handState === "REACHING") return true;
  if (handState === "GRIPPED" && triggerValue >= 0.5) return true;
  return false;
}

function anyHandGripped(actor: ActorState): boolean {
  return actor.leftHand.state === "GRIPPED" || actor.rightHand.state === "GRIPPED";
}

function allLimbsIdle(actor: ActorState): boolean {
  return (
    actor.leftHand.state === "IDLE" &&
    actor.rightHand.state === "IDLE" &&
    actor.leftFoot.state === "LOCKED" &&
    actor.rightFoot.state === "LOCKED"
  );
}

// -----------------------------------------------------------------------------

export function updateStamina(
  prev: number,
  inputs: StaminaInputs,
  cfg: StaminaConfig = DEFAULT_STAMINA_CONFIG,
): number {
  const dtSec = inputs.dtMs / 1000;
  let ratePerSec = 0;

  // Drain — hands active.
  if (handIsActive(inputs.actor.leftHand.state, inputs.triggerL) ||
      handIsActive(inputs.actor.rightHand.state, inputs.triggerR)) {
    ratePerSec -= cfg.handActiveDrainPerSec;
  }

  // Drain — sustained posture break (scaled by magnitude).
  // §5.2 phrases this as "攻め側" so we read the TOP actor's break from
  // the caller's `actor` only if this stamina update is for the top;
  // bottom actor's own break is what drains THEM. Stage 1 drives only
  // bottom stamina right now, so we read bottom.postureBreak — which is
  // zero until top-side wiring lands. The code is future-proof for the
  // day defender stamina is wired.
  const breakMag = breakMagnitude(inputs.actor.postureBreak);
  if (breakMag > 0) {
    ratePerSec -= cfg.postureMaintainDrainPerSec * breakMag;
  }

  // Recovery — BTN_BREATH + no grips + static hip.
  const hipMag = Math.hypot(inputs.attackerHip.hip_push, inputs.attackerHip.hip_lateral);
  const staticHip = hipMag < cfg.hipInputStaticThreshold;
  if (inputs.breathPressed && !anyHandGripped(inputs.actor) && staticHip) {
    ratePerSec += cfg.breathRecoverPerSec;
  } else if (allLimbsIdle(inputs.actor) && !anyHandGripped(inputs.actor)) {
    // Idle recovery only applies when nothing is happening. Breath
    // recovery takes priority because it's explicit player intent.
    ratePerSec += cfg.idleRecoverPerSec;
  }

  const next = prev + ratePerSec * dtSec;
  return clamp01(next);
}

// §5.2 last row — confirming a technique deducts a flat cost.
export function applyConfirmCost(
  prev: number,
  cfg: StaminaConfig = DEFAULT_STAMINA_CONFIG,
): number {
  return clamp01(prev - cfg.confirmDrainFlat);
}

// §5.3 — clamp grip strength ceiling at low stamina. Returns a ceiling in
// [0,1] that downstream should apply as `min(raw_trigger, ceiling)`.
export function gripStrengthCeiling(
  stamina: number,
  cfg: StaminaConfig = DEFAULT_STAMINA_CONFIG,
): number {
  if (stamina < cfg.lowGripCapThreshold) return cfg.lowGripCap;
  return 1;
}

// §5.3 — below this stamina, new REACHING is refused.
export function canStartReach(
  stamina: number,
  cfg: StaminaConfig = DEFAULT_STAMINA_CONFIG,
): boolean {
  return stamina >= cfg.noReachThreshold;
}

function clamp01(x: number): number {
  if (x < 0) return 0;
  if (x > 1) return 1;
  return x;
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/State/Stamina.cs
// Ported 1:1 from src/prototype/web/src/state/stamina.ts.
// PURE — stamina update per docs/design/state_machines_v1.md §5.

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // Config (§5.2 rate table)
    // -------------------------------------------------------------------------

    public struct StaminaConfig
    {
        // §5.2
        public float HandActiveDrainPerSec;      // any hand REACHING / GRIPPED(≥0.5)
        public float PostureMaintainDrainPerSec; // × ‖posture_break‖
        public float ConfirmDrainFlat;           // applied once per confirm
        public float BreathRecoverPerSec;        // BTN_BREATH + no grips + static
        public float IdleRecoverPerSec;          // all IDLE, no grips

        public float HipInputStaticThreshold;    // §5.2 "hip_input < 0.2"

        // §5.3
        public float LowGripCap;                 // grip strength clamp at low stamina
        public float LowGripCapThreshold;        // stamina level below which cap activates
        public float NoReachThreshold;           // stamina below which REACHING is refused

        public static readonly StaminaConfig Default = new StaminaConfig
        {
            HandActiveDrainPerSec      = 0.02f,
            PostureMaintainDrainPerSec = 0.05f,
            ConfirmDrainFlat           = 0.1f,
            BreathRecoverPerSec        = 0.1f,
            IdleRecoverPerSec          = 0.03f,
            HipInputStaticThreshold    = 0.2f,
            LowGripCap                 = 0.6f,
            LowGripCapThreshold        = 0.2f,
            NoReachThreshold           = 0.05f,
        };
    }

    // -------------------------------------------------------------------------
    // Attacker (bottom) stamina inputs
    // -------------------------------------------------------------------------

    public struct StaminaInputs
    {
        public float     DtMs;
        public ActorState Actor;
        public HipIntent AttackerHip;
        public float     TriggerL;
        public float     TriggerR;
        public bool      BreathPressed; // BTN_BREATH held or edge this frame
    }

    // -------------------------------------------------------------------------
    // Defender (top) stamina inputs
    // -------------------------------------------------------------------------

    public struct StaminaDefenderInputs
    {
        public float     DtMs;
        public ActorState Actor;         // the top actor
        public float     LeftBasePressure;
        public float     RightBasePressure;
        public float     WeightForward;
        public float     WeightLateral;
        public bool      BreathPressed;
    }

    // -------------------------------------------------------------------------
    // Pure operations
    // -------------------------------------------------------------------------

    public static class StaminaOps
    {
        public static float UpdateStamina(
            float prev,
            StaminaInputs inp,
            StaminaConfig cfg = default)
        {
            if (cfg.BreathRecoverPerSec == 0f) cfg = StaminaConfig.Default;

            float dtSec = inp.DtMs / 1000f;
            float rate  = 0f;

            // Drain — hands active.
            if (HandIsActive(inp.Actor.LeftHand.State, inp.TriggerL) ||
                HandIsActive(inp.Actor.RightHand.State, inp.TriggerR))
            {
                rate -= cfg.HandActiveDrainPerSec;
            }

            // Drain — sustained posture break.
            float breakMag = PostureBreakOps.BreakMagnitude(inp.Actor.PostureBreak);
            if (breakMag > 0f)
                rate -= cfg.PostureMaintainDrainPerSec * breakMag;

            // Recovery — BTN_BREATH + no grips + static hip.
            float hipMag    = (float)System.Math.Sqrt(inp.AttackerHip.HipPush * inp.AttackerHip.HipPush + inp.AttackerHip.HipLateral * inp.AttackerHip.HipLateral);
            bool  staticHip = hipMag < cfg.HipInputStaticThreshold;

            if (inp.BreathPressed && !AnyHandGripped(inp.Actor) && staticHip)
            {
                rate += cfg.BreathRecoverPerSec;
            }
            else if (AllLimbsIdle(inp.Actor) && !AnyHandGripped(inp.Actor) && staticHip)
            {
                rate += cfg.IdleRecoverPerSec;
            }

            return Clamp01(prev + rate * dtSec);
        }

        public static float UpdateStaminaDefender(
            float prev,
            StaminaDefenderInputs inp,
            StaminaConfig cfg = default)
        {
            if (cfg.BreathRecoverPerSec == 0f) cfg = StaminaConfig.Default;

            float dtSec = inp.DtMs / 1000f;
            float rate  = 0f;

            bool leftActive  = inp.LeftBasePressure  >= 0.5f;
            bool rightActive = inp.RightBasePressure >= 0.5f;
            if (leftActive || rightActive)
                rate -= cfg.HandActiveDrainPerSec;

            float breakMag = (float)System.Math.Sqrt(inp.Actor.PostureBreak.X * inp.Actor.PostureBreak.X + inp.Actor.PostureBreak.Y * inp.Actor.PostureBreak.Y);
            if (breakMag > 0f)
                rate -= cfg.PostureMaintainDrainPerSec * breakMag;

            float weightMag   = (float)System.Math.Sqrt(inp.WeightForward * inp.WeightForward + inp.WeightLateral * inp.WeightLateral);
            bool  staticWeight = weightMag < cfg.HipInputStaticThreshold;
            bool  anyBaseActive = inp.LeftBasePressure > 0f || inp.RightBasePressure > 0f;

            if (inp.BreathPressed && !anyBaseActive && staticWeight)
            {
                rate += cfg.BreathRecoverPerSec;
            }
            else if (!anyBaseActive && staticWeight && AllLimbsIdle(inp.Actor))
            {
                rate += cfg.IdleRecoverPerSec;
            }

            return Clamp01(prev + rate * dtSec);
        }

        // §5.2 last row — confirming a technique deducts a flat cost.
        public static float ApplyConfirmCost(float prev, StaminaConfig cfg = default)
        {
            if (cfg.ConfirmDrainFlat == 0f) cfg = StaminaConfig.Default;
            return Clamp01(prev - cfg.ConfirmDrainFlat);
        }

        // §5.3 — grip-strength ceiling at low stamina.
        public static float GripStrengthCeiling(float stamina, StaminaConfig cfg = default)
        {
            if (cfg.LowGripCap == 0f) cfg = StaminaConfig.Default;
            return stamina < cfg.LowGripCapThreshold ? cfg.LowGripCap : 1f;
        }

        // §5.3 — below this stamina new REACHING is refused.
        public static bool CanStartReach(float stamina, StaminaConfig cfg = default)
        {
            if (cfg.NoReachThreshold == 0f) cfg = StaminaConfig.Default;
            return stamina >= cfg.NoReachThreshold;
        }

        // --- classification helpers -------------------------------------------

        static bool HandIsActive(HandState state, float triggerValue)
        {
            if (state == HandState.Reaching) return true;
            if (state == HandState.Gripped && triggerValue >= 0.5f) return true;
            return false;
        }

        static bool AnyHandGripped(ActorState actor) =>
            actor.LeftHand.State == HandState.Gripped || actor.RightHand.State == HandState.Gripped;

        static bool AllLimbsIdle(ActorState actor) =>
            actor.LeftHand.State  == HandState.Idle &&
            actor.RightHand.State == HandState.Idle &&
            actor.LeftFoot.State  == FootState.Locked &&
            actor.RightFoot.State == FootState.Locked;

        static float Clamp01(float v) => v < 0f ? 0f : v > 1f ? 1f : v;
    }
}
